### ISSUE TITLE
pause/resume with enable_cpu_backup=True corrupts model weights

### ISSUE THREAD

**Issue by Heronalps** [fzyzcjy/torch_memory_saver#71]
## Summary

`pause()` + `resume()` with `enable_cpu_backup=True` does not correctly restore GPU tensor contents. After a pause/resume cycle, model weights are corrupted, producing completely different (garbage) inference output compared to the same model without pause/resume.

## Environment

- `torch_memory_saver==0.0.9`
- PyTorch 2.5.1
- CUDA (H100 80GB)
- Python 3.12
- Running inside Modal containers (but issue reproduces without GPU snapshots — see ablation below)

## Reproduction

We have a GPT language model (~6B params) that we load inside a `region("model", enable_cpu_backup=True)`, with a KV cache in `region("kvcache", enable_cpu_backup=False)`. We then call `pause()` and `resume()` and run inference.

```python
import torch_memory_saver

# Engine subprocess sets up regions during model loading:
with torch_memory_saver.region("model", enable_cpu_backup=True):
    model = load_model(checkpoint_path)  # ~6B param GPT model

with torch_memory_saver.region("kvcache", enable_cpu_backup=False):
    kv_cache = model.setup_caches(block_size=256)

# Later: pause to free GPU memory
torch_memory_saver.pause("kvcache")
torch_memory_saver.pause("model")

# Resume to restore
torch_memory_saver.resume("kvcache")
torch_memory_saver.resume("model")

# Run inference — output is corrupted
output = model.generate(prompt, seed=42)
# Expected: ~2000+ tokens of coherent audio semantic codes
# Actual: ~100 tokens of garbage before hitting EOS
```

## Ablation results

We ran a systematic ablation to isolate the root cause. All tests use the same model, same seed (42), same prompt:

| Test | `region()` wrapping | `pause/resume` | Output length | First tokens correct? |
|------|---------------------|----------------|---------------|-----------------------|
| Baseline (no torch_memory_saver) | No | No | 1938–2374 | ✅ Yes |
| Region only, no pause/resume | Yes | **No** | 1873–2373 | ✅ Yes |
| Region + pause/resume | Yes | **Yes** | 100–287 | ❌ No |

Key findings:
- **Row 2 vs Row 1**: Wrapping model loading in `region()` alone does NOT affect output — model works correctly.
- **Row 3 vs Row 2**: Adding `pause()` + `resume()` corrupts the output. Sequences are 10–20x shorter and tokens are completely different from the first position.
- The corruption is **deterministic** — running the same pause/resume test multiple times produces identical (but wrong) output.

## Additional context

- We tested both `hook_mode="torch"` and `hook_mode="preload"` (with `configure_subprocess()`) — same corruption in both modes.
- We also tested with and without `torch.compile(mode="reduce-overhead")` — same corruption regardless. This rules out CUDA graph interaction.
- The issue was initially discovered in a Modal GPU snapshot context, but the ablation above proves it reproduces with plain pause/resume, no GPU snapshots involved.
- The model uses `forkserver` multiprocessing — the engine runs in a child process. `configure_subprocess()` is used when spawning.

## Expected behavior

After `pause("model")` + `resume("model")` with `enable_cpu_backup=True`, the GPU tensor contents should be identical to before the pause. The README states:

> If you want the tensor content to be kept unchanged, use `enable_cpu_backup`.
> ```python
> assert tensor1[0] == 42, "content is kept unchanged"
> ```

## Actual behavior

Tensor contents are corrupted after resume. The model produces completely different (wrong) output from the very first inference step, indicating the weights themselves are wrong.

**Comment by Heronalps** [comment]
Potentially related: #61 — reports that pause doesn't release memory for CUDA graphs. Our issue may share a similar root cause where the pause/resume mechanism doesn't properly handle certain allocation patterns.

**Comment by fzyzcjy** [comment]
Hmm could you please share a minimal reproducible sample? Then I can have a check about it.

**Comment by DoradusResearch** [comment]
+1 — happy to provide a minimal repro from a production setup. We run a 3-peer TP=4 rotation pool on consumer Blackwell SM12.0a (4× RTX PRO 6000) under `vllm/vllm-openai:v0.20.2-cu129-ubuntu2404` with `--enable-sleep-mode --enable-weights-cpu-backup` (which routes through torch_memory_saver with `enable_cpu_backup=True`). Our cycle telemetry below trips weight corruption inside ~4 rotation cycles:

**Concrete leak telemetry from a production 3-peer TP=4 rotation pool** (AB1, 4× RTX PRO 6000 Blackwell SM12.0a, vllm/vllm-openai:v0.20.2-cu129-ubuntu2404 base). Three vLLM allocs (GLM-5.1-REAP, MiMo-V2.5, DSv4-Flash) share GPUs 0–3; only one is `/wake_up` at a time, others stay `/sleep` level=1. Hot-loop /sleep + /wake_up across the three peers under `ab1-tp4-rotate.sh`, tracing the CuMemAllocator per cycle:

| Cycle | Freed on /sleep | Still pinned post-sleep | Delta vs cycle 1 |
|---|---|---|---|
| 1 | 25.3 GiB | 13.2 GiB | baseline |
| 2 | 25.6 GiB | 16.6 GiB | +3.4 GiB |
| 3 | 25.6 GiB | 24.1 GiB | +7.5 GiB |
| 4 | 25.6 GiB | 68.7 GiB | +44.5 GiB |

By cycle 4 the leaked allocator pages overlap the weight tensors and corrupt them in place. HTTP 200 throughout, `/is_sleeping` reports `true`, nothing in stderr — but subsequent inference produces uniform garbage (all top-5 logprobs identical at -11.9504, output `!!!!` repeated). Recovery requires full alloc restart; `/sleep` level=2 was tried and made it worse (corruption observable after a single cycle, matches #29341).

**Minimal reproduction recipe** (3 peers, TP=4, single host):

1. Start three vllm allocs sharing the same 4 GPUs, each with `--enable-sleep-mode --enable-weights-cpu-backup` and `--gpu-memory-utilization 0.85`. Only one peer is awake at a time; others were `/sleep` level=1 immediately after their initial weight load.
2. In a loop, every ~10 min, pick the next peer in round-robin and:
   - `POST <awake_peer>/sleep?level=1` (wait for `/is_sleeping=true`)
   - wait 35s (post-sleep gap — required on jasl/cu130 base, optional on v0.20.2)
   - `POST <next_peer>/wake_up`
3. After each cycle, smoke-test with a chat completion against the now-awake peer (`max_tokens=10, temperature=0`).
4. By cycle 4 outputs degenerate: all top-5 logprobs identical at -11.9504, body is `!!!!` repeated. HTTP 200 throughout, no exception in stderr.

Cross-posted to vllm-project/vllm #29341 / #34600 / #35489 / #36651 / #37111 / #21336 — torch_memory_saver appears to be the underlying primitive in several of those reports, hence the +1 here.

Filing from DoradusResearch.

**Comment by zwc163** [comment]
I meet the same problem in verl. Even though I set enable_draft_weights_cpu_backup=True, after resuming, the MTP weights are still a tensor of all zeros

**Comment by fzyzcjy** [comment]
Hi @zwc163 could you please share a reproducible sample
